analyzer/emerald/incoming.go

```diff
@@ -400,7 +432,311 @@ func extractRound(b *block.Block, txrs []*sdkClient.TransactionWithResults, logg
+    },
+    {
+      "constant": true,
+      "inputs": [],
+      "name": "symbol",
+      "outputs": [
+        {
+          "name": "",
+          "type": "string"
+        }
+      ],
+      "payable": false,
+      "stateMutability": "view",
+      "type": "function"
+    },
+    {
+      "constant": false,
+      "inputs": [
+        {
+          "name": "_to",
+          "type": "address"
+        },
+        {
+          "name": "_value",
+          "type": "uint256"
+        }
+      ],
+      "name": "transfer",
+      "outputs": [
+        {
+          "name": "success",
+          "type": "bool"
+        }
+      ],
+      "payable": false,
+      "stateMutability": "nonpayable",
+      "type": "function"
+    },
+    {
+      "constant": true,
+      "inputs": [
+        {
+          "name": "_owner",
+          "type": "address"
+        },
+        {
+          "name": "_spender",
+          "type": "address"
+        }
+      ],
+      "name": "allowance",
+      "outputs": [
+        {
+          "name": "remaining",
+          "type": "uint256"
+        }
+      ],
+      "payable": false,
+      "stateMutability": "view",
+      "type": "function"
+    },
+    {
+      "anonymous": false,
+      "inputs": [
+        {
+          "indexed": true,
+          "name": "_from",
+          "type": "address"
+        },
+        {
+          "indexed": true,
+          "name": "_to",
+          "type": "address"
+        },
+        {
+          "indexed": false,
+          "name": "_value",
+          "type": "uint256"
+        }
+      ],
+      "name": "Transfer",
+      "type": "event"
+    },
+    {
+      "anonymous": false,
+      "inputs": [
+        {
+          "indexed": true,
+          "name": "_owner",
+          "type": "address"
+        },
+        {
+          "indexed": true,
+          "name": "_spender",
+          "type": "address"
+        },
+        {
+          "indexed": false,
+          "name": "_value",
+          "type": "uint256"
+        }
+      ],
+      "name": "Approval",
+      "type": "event"
+    }
+  ]
+`))
+	if err != nil {
+		panic(err)
+	}
+}
+
+// nolint: unparam
+func evmCallWithABI(
+	ctx context.Context,
+	client *oasis.RuntimeClient,
+	round uint64,
+	contractEthAddr []byte,
+	contractABI *abi.ABI,
+	name string,
+	outI interface{},
+	in ...interface{},
+) error {
+	// https://github.com/oasisprotocol/emerald-web3-gateway/blob/v3.0.0/rpc/eth/api.go#L408
+	gasPrice := []byte{1}
+	gasLimit := uint64(30_000_000)
+	caller := ethCommon.Address{1}.Bytes()
+	value := []byte{0}
+
+	inPacked, err := contractABI.Pack(name, in...)
+	if err != nil {
+		panic(err)
+	}
+	outPacked, err := client.EvmSimulateCall(ctx, round, gasPrice, gasLimit, caller, contractEthAddr, value, inPacked)
+	if err != nil {
+		return fmt.Errorf("runtime client evm simulate call: %w", err)
+	}
+	if err = contractABI.UnpackIntoInterface(outI, name, outPacked); err != nil {
+		return fmt.Errorf("unpack out: %w", err)
```

Comment:
gonna have to write a longer response to this :skull:

Reply:
my opinion when it comes to creating errors, wrapping errors, and handling errors

when wrapping an error in another error, only provide context. we _don't_ want error messages like "failed to do A: error in B: failed to do C: error in D." it's excessive to say something failed multiple times. instead, prefer to build error messages like "scan failed: in A: in B: in C: in D condition"

- the innermost error ought to have a condition stated as fact, with no interpretation about its supposed impact on the program
  - good: `fmt.Errorf("cache does not contain key %s", key)`
  - bad, 'crashing' inappropriate: `fmt.Errorf("crashing due to missing key %s", key)`
- when wrapping errors, provide context. identify where the error came from. again, don't include interpretation about its supposed impact on the program
  - good, in a function that looks up sender and receiver: `fmt.Errorf("looking up sender: %w", err)`
  - bad, missing context about which lookup failed: `fmt.Errorf("cache lookup: %w", err)`
  - good, in a function that loops through a transaction list: `fmt.Errorf("transaction %d getting affected accounts: %w", i, err)`
  - bad, can find source code line, but not iteration in loop: `fmt.Errorf("getting affected accounts: %w", err)`
  - bad, 'failed' inappropriate: `fmt.Errorf("failed to look up cached sender: %w", err)`
  - bad, throwing away error info: `fmt.Errorf("failed to look up cached sender", txHash)`
  - fine, in a straight-line function that only calls another function, so no additional context is needed: `err`
- when handling the error i.e. on the outermost layer, describe the impact to the program
  - good: `logger.Error("unable to process transaction, skipping", "err", err)`
  - bad, no description of impact: `logger.Error("processing transaction", "err", err)`
  - bad, throwing away error info `logger.Error("unable to process transaction, skipping")`
  - good, as panic already says that it is panicking: `panic(fmt.Sprintf("parsing hardcoded special address %s: %v", SpecialAddress, err))`

if A calls B, and B fails, who should say it was B? imo, A

- (i) "parsing To address: checksum mismatch" vs (ii) "To: parsing address checksum mismatch" I like (i)
- our codebase tends to have fewer callsites for each function and more places in each function that returns with error 

how detailed should the context be when wrapping:

- at a minimum, it should uniquely identify a location in the function, otherwise can be unclear what's failing
- when the control flow is complex, add information about that, e.g. iteration in a loop
- as mentioned by @mitjat, it can be helpful if it is unique in the whole project, so people can arrive by searching
- try to keep the context about what's inside the function, so that you don't need to adjust much when code paths leading to the function change

Reply:
so as it applies to this line:

- ![image](https://user-images.githubusercontent.com/38229038/206811697-cd342ce5-ba6b-4c25-93b1-6444adefabb2.png)
 by some luck, this really is unique in the codebase
- "unpack" is enough to identify UnpackIntoInterface as the source of the error
- incidentally there's only one call to UnpackIntoInterface, but specifying `out` might be helpful if later refactoring moves this code to somewhere with other similar calls
- identifying that it's "evm contract call" we should do in the caller of `evmCallWithABI`. currently that's only "ERC20 call failed." I don't mind adding that it's an EVM contract call there
- "failed to" I'm avoiding because the code that handles the error should already indicate that something failed
- the string here is short, so there's the fear that it won't be unique. but I don't want us to choose other conventions to make it longer. if we follow different conventions, it's about the same amount of information and would have the same likelihood of repeating in the codebase:
  - include "failed to": "unpack out" -> "failed to unpack out"
  - callee identifies itself instead of caller identifies callee: "failed to unpack out" -> "evm call failed, out"
- identifying the callee in both the caller an callee I think would give us a better chance at uniqueness, but it's repetitive in the resulting error message: "download round tokens downloading erc20: download erc20 making evm call for name: evm call for name unpacking out" :shrug:

Reply:
lengthened this and a few others

Reply:
I think "unpack out" made me comment because it's so ungrammatical / hard to parse without context. It wasn't clear that "out" is a variable name and not a weird preposition. I guess I touched on something :). Anyway, I like what you landed on. Thank you :) 

I agree with your guidelines almost entirely. The only slight difference in opinion is that in an ideal world, a function is self-contained and should not / cannot know or care whether it's the the outermost layer of error message compositing. If you subscribe to that idea, it encourages more verbosity at the cost of some repetition in the final composed message (which I personally also don't mind much).

The last comment made me laugh. I emotionally relate to the process :)
"Somebody is Wrong on the Internet™" -> over-engage -> "ya alright let's do it your way, let's wrap this up, it doesn't matter much"